### ISSUE TITLE
Print() method - show line number and node where the print method was emited from

### ISSUE THREAD

**Issue by blurymind** [godotengine/godot#18319]
The print method is a great debugging tool, but sometimes when we have it at many places of a script - it becomes somewhat difficult to trace back where it came from.

It would immencely help for debugging if when printing in godot editor, it also showed the node and line number of the script it got called from

**Comment by groud** [comment]
What for ? This is going to need a complex code instrumentation for you to retrieve a debugging text you wrote yourself ?
For debugging complex things we have a debugger. Printing too many stuff to debug is never a good idea.

**Comment by karroffel** [comment]
For GDNative there are error and warning print methods, those show the file and line number where the print came from. These prints will display in Godot like errors and warning emitted from the engine itself.

https://github.com/godotengine/godot/blob/7d6f210ccb5de9ef414f94ad42f9f3dea14c0493/modules/gdnative/gdnative/gdnative.cpp#L131-L141

Only a few print types are supported yet

https://github.com/godotengine/godot/blob/8fc298be5f859fdb9b7b228793554dee6b025a09/core/error_macros.h#L51-L56

Maybe an `ERR_HANDLER_INFO` could be added for that purpose?

**Comment by groud** [comment]
Having a dedicated error function might be a good idea though, for this I guess it would make sense.

**Comment by blurymind** [comment]
@groud  How are we going to call it?
printd?

**Comment by Paulb23** [comment]
Related to #3884 and  #13335

**Comment by groud** [comment]
perror() maybe?

**Comment by Noshyaar** [comment]
`perror` is ambiguous imo. It's not obvious what `p` stands for, and might get confused with `print_err`

#3884 suggests `printl` which I think is alright, but by existing methods `prints` and `printt` put separator between parameters. It could be misunderstood as such method because of its name.

I suggest `printdbg` or `printdebug` since we seem to prefer longer name than short-but-ambiguous name.

**Comment by Zylann** [comment]
You could make your own log function using print_stack(), but it would include the custom function too.

**Comment by Zylann** [comment]
In fact, stack trace only makes sense for errors. For debugging, you have the debugger and print_stack(). To clean up your prints, either prefix them with "DDD" or something, or globsearch for 'print'.

**Comment by chanon** [comment]
OK, I created a pull request with this implemented.
The method name is currently `printdbg`, but I still prefer `trace`.
The output is displayed as
```
Printed message		from res://path/to/file.gd:34 in function '_ready'`
```

**Comment by chanon** [comment]
After trying this out in my project, I found the output is a bit hard to parse. 

Here is an example:
![image](https://user-images.githubusercontent.com/357737/40177785-bd6db112-5a09-11e8-9081-a3a43c0abbd3.png)

It looks messy.

I tried putting the source on a different line and I think it is easier to read, however it creates more lines.
![image](https://user-images.githubusercontent.com/357737/40177876-00d3c4e6-5a0a-11e8-8b9e-2cee0639e097.png)

**Comment by Zylann** [comment]
Multiline is how Unity does it, however it also separates all prints by a newline.

I also remembered that non-blocking breakpoints can also be useful to fulfill some debug print use cases https://github.com/godotengine/godot/issues/6576

**Comment by chanon** [comment]
Here is the same print output without the 'in function...' part and using 'At: ' as a prefix like in the engine error log

![image](https://user-images.githubusercontent.com/357737/40178258-311ba08c-5a0b-11e8-8c2c-f758529c6580.png)

**Comment by chanon** [comment]
What could also be done is to expose the stack information that print_stack uses to GDScript, so people could create their own logging method in GDScript. I'm not sure how/where to do that though. But if anyone is interested, you can look at how PRINT_STACK is implemented in `gdscript_functions.cpp`. With that info you could create your own custom print function in a C++ module and format however you want (which is what I'm going to be doing actually).

**Comment by blurymind** [comment]
@chanon your method looks great. It would be nice if it's a part of gdevelop imo - its a pretty fundamental debugging feature

**Comment by akien-mga** [comment]
> It would be nice if it's a part of gdevelop imo - its a pretty fundamental debugging feature

Wrong FOSS engine :wink:

**Comment by blurymind** [comment]
filed another feature request for the searh filter:
https://github.com/godotengine/godot/issues/19098

In any case I should add that it would also be useful if the script name and line output are also clickable. When clicked it should open the script and line it points to.

**Comment by Zylann** [comment]
Do we have an issue open about logging in general? I've seen quite a lot of issues relating to this topic so far.

**Comment by blurymind** [comment]
I also raised an issue request for error stacking here
https://github.com/godotengine/godot/issues/19121

@Zylann I dont think there is one - but it would be nice to have one :) 
Would you like me to open it?

**Comment by chanon** [comment]
With #18966 and #18976 merged, I think this can be closed.

**Comment by blurymind** [comment]
@chanon 🥇 Thank you =)

**Comment by Zylann** [comment]
And if we do https://github.com/godotengine/godot/issues/19974 we will be able to track normal prints too, I guess?

**Comment by TheConceptBoy** [comment]
I am all for adding this even to 3.x because when you get in the zone, you just add print commands for debugging and checking and not always remove them right away. You can use search to find print commands but if you tend to comment out prints for later use you will also see them all crop up as numerous lines of 
```
#print()
#print()
#print()
#print()
print()
#print()
#print()
```
 
If the console displays the line number, you can simply click on that line and it will take you to the script and line. It's unbelievable convenient and there's a very good reason why Web Dev tools have had them for ages. Just like Godot scripts, Web Dev scripts can potentially get ginormous. Godot scripts, even with inheritance, can get huge. There can and will be dozens of prints commands being used and left behind while you're in the zone of scripting. The whole idea is to help you locate the print calls.

![image](https://user-images.githubusercontent.com/52581279/174495946-a2c8d1d8-a4fe-4ac3-82f8-3a28fe13fafe.png)

**Comment by Calinou** [comment]
@TheConceptBoy You can use `print_debug()` for this purpose. Doing this for all `prints()` unconditionally would be too noisy, and that can be counterproductive to debugging too.

**Comment by zydeas** [comment]
This would be very useful and I'm not sure why people had any issue at all with this when it was suggested? Literally right now I am dealing with an issue where I stuck a print statement somewhere that's spamming every frame during a specific mode in my game, and even by doing text searches for function calls I can't find the specific statement that's causing it, because the pattern for what is being printed out is non-standard. I would save so much time by just being able to click or hover over a debug line and see what code path it was called from. You could do it without any visual noise at all! There's a reason Unity has this feature: Because it's useful!

**Comment by AThousandShips** [comment]
> This would be very useful and I'm not sure why people had any issue at all with this when it was suggested?

It was closed because it was implemented, this was never rejected, it was completed

**Comment by zydeas** [comment]
Ahh right, sorry, didn't realize this was the old proposal method. I just got here from google.

**Comment by Calinou** [comment]
There's already a proposal about making `print()` output clickable: https://github.com/godotengine/godot-proposals/issues/9367
